docs/understanding-airbyte/jobs.md

```diff
@@ -1,45 +1,111 @@
+The source process should automatically exit after passing all of its messages. Similarly, the destination process shutdowns after receiving all records. Each process is given a shutdown grace period. The worker forces shutdown if this is exceeded.
 
 ![Worker Lifecycle](../.gitbook/assets/worker-lifecycle.png)
 
 [Image Source](https://docs.google.com/drawings/d/1k4v_m2M5o2UUoNlYM7mwtZicRkQgoGLgb3eTOVH8QFo/edit)
 
 See the [architecture overview](high-level-view.md) for more information about workers.
 
-## Worker parallelization
+## Deployment Types
+
+Up to now, the term 'processes' has been used loosely. This section will describe this in more detail.
+
+Airbyte offers two deployment types. The underlying process implementations differ accordingly.
+
+1. The Docker deployment - Each process is a local process backed by a Docker container. As all processes are local, process communication is per standard unix pipes.
+2. The Kubernetes deployment - Each process is a backed by a Kubernetes pod. As Kubernetes does not make process-locality guarantees, Airbyte has implemented mechanisms to hide the remote process execution.
+   See [this blogpost](https://airbyte.com/blog/scaling-data-pipelines-kubernetes) for more details.
+
+### Decoupling Worker and Job Processes
+
+Workers being responsible for all non-connector-related job operations means multiple jobs are operationally dependent on a single worker process.
+
+There are two downsides to this:
+1. Any issues to the parent worker process affects all job processes launched by the worker.
+2. Unnecessary complexity of vertically scaling the worker process to deal with IO and processing requirements from multiple jobs.
+
+This gives us a potentially brittle system component that can be operationally tricky to manage. For example, since redeploying Airbyte terminates all worker processes, all running jobs are also terminated.
+
+The Container Orchestrator was introduced to solve this.
+
+#### Container Orchestrator
+When enabled, workers launch the Container Orchestrator process.
+
+This orchestrator process replaces the previous worker process. It now performs the [above listed responsibilities](#worker-responsibilities).
+
+This decoupling introduces a new need for workers to track the orchestrator's, and the job's, state. This is done via a shared Cloud Storage store.
+
+Brief description of how this works,
+1. Workers constantly poll the Cloud Storage location for job state.
+2. As an Orchestrator process executes, it writes status marker files to the Cloud Storage location i.e. `NOT_STARTED`, `INITIALIZING`, `RUNNING`, `SUCCESS`, `FAILURE`.
+3. If the Orchestrator process runs into issues at any point, it writes a `FAILURE`.
```

Comment:
Do we have a process today to detect jobs that crashed before writing `FAILURE`?

Reply:
This is a potential hole today. 

@pmossman and I were just debugging a sync where this happened. This is the first time I've seen this happen though definitely maybe because it should be happening more often.

https://github.com/airbytehq/airbyte/blob/master/airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/ContainerOrchestratorApp.java#L129 is where we catch the throwable. Very possible something is interfering and this doesn't hit.

Maybe the LauncherWorker is a good place to add this.

Reply:
Now that we are adding back some garbage collection like job (https://github.com/airbytehq/airbyte/issues/14043), it might make sense to have something for this use case.
We should probably create a story to track this, I suspect it might be more complex because with multi cloud, that will involve finding orphaned jobs in the data planes. Also, GCS might not be super happy with a scan on all the states.

Reply:
Done: https://github.com/airbytehq/airbyte/issues/16665.

I think we should create a story to investigate the cases first. Once we know why, we can propose fixes. Want to make sure we properly fix this and don't act too hasty.

Reply:
Absolutely, we need to know what we're fixing